### PR TITLE
Synchronous communication in EDA

### PR DESCRIPTION
**Pull request by azarboon**
Before clicking the "Create pull request" button:

Enter a meaningful title. We suggest using the keywords "New" or "Update" indicating the nature of changes.

#### Summary, scope, and intent of this PR

Sometimes, it's needed to have synchronous communications within EDA. I have added two ways  of implementing it.  I've made this PR following [my chat](https://github.com/MicrosoftDocs/architecture-center/pull/4575#issuecomment-2444546700) with @ckittel

#### Links to any related work item(s) or supporting detail

- [DELETE-OR-REPLACE-THIS-TEXT]
- [DELETE-OR-REPLACE-THIS-TEXT]

  

<details><summary>AFTER YOUR PR HAS BEEN CREATED, expand this section for tips and additional instructions.</summary>
These are common guidelines for contributions across the repos managed by the Azure patterns & practices (PnP) content team. Some repositories may have additional specific requirements that are not listed here.   

## Guidance for all contributors  
  
| Topic | Guidance |
| ----------| ------------ |
| **Draft PR** | If your PR will be a work-in-progress for more than a day or two, select the **Convert to draft** link in the upper right of the page (under **Reviewers**) to change it to a draft. For future reference, you can also do this using the **Create pull request** button drop-down during PR creation. | 
| **ms.date metadata** | <ul><li>Don't update an article's "ms.date" metadata property unless you've done a **full freshness review** of the content. A full freshness review includes changes required to correct or improve the **full** technical accuracy of the article.</li><li>Don't update "ms.date" if you're doing targeted changes to improve non-technical aspects of the article, such as editorial quality, art improvements, article template alignment, etc.</li><li>If you've changed any "ms.date" properties for work that wasn't part of full review for freshness, please reset them to their previous value.</li></ul> | 
| **Placement and linking** | If you're creating a new article or articles, include updates to the related TOC.yml file to propose where the article(s) should be placed. Also consider other places within the document set where it would be beneficial to cross-reference and link to your new article(s). | 
| **PR build** | After you open your PR, and for each successive commit that you push to your branch, the publishing platform will run validation on the files in your pull request. A summary of the build results for each file will be inserted inline into your pull request, which includes any build suggestions/warnings/errors. PRs cannot be merged until all build errors and most warnings are resolved. |
| **Publishing** | Following a successful merge, most repos publish to the live site about once per (business) day. |
| **Additional resources** | <ul><li>[learn.microsoft.com contributor guide](https://review.learn.microsoft.com/help/contribute/?branch=main)</li></ul>
  
## Additional guidance for private repos and internal contributors  

| Topic | Guidance | 
| ----------| ------------ | 
| **PR size** | If your PR is more than ~5 lines of changes, or you'd like for the changes to go through editorial or larger review, open a contribution request at <https://aka.ms/contribution> and include a link to the PR in response #8. Once it's processed, you'll be notified of the next steps.  |
| **PR title prefix** | Select the **Edit** button to the right of the PR title if you need to revise it. The following prefixes are reserved for specific contribution types:<br/><br/><ul><li>**[Quality Check]** - maintenance work related to content quality (edit passes, art improvements, template alignment)</li><li>**[LinkFix]** - recurring/adhoc PRs to correct link URLs</li><li>**[Pipeline]** - new/updated contributor success pipeline content</li><li>**[WIP]** - a work-in-progress draft requiring several days/weeks</li></ul> |
| **PR preview** | Following successful build of your PR, publishable files will also include **Preview URL** links to staged previews of your new/updated articles. Be sure to review these for verification of your intended contributions, or to send to other internal contributors for review. |
| **PR sign-off (public repo)** | If an article you own is updated in a public repo PR, you are responsible for sign-off. You will be automatically notified via email. The PR will not be merged until you've had a chance to review and sign-off. |
| **PR sign-off (private repo)** | After you've completed your proposed changes, addressed build warnings, and completed all review work, you can begin the sign-off process for review and merge:<br/><br/><ol><li>If your PR is in draft mode, remove "[WIP]" from the title and select **Ready for review** button at the bottom of the PR.</li><li>Enter "#sign-off" in a new comment. This comment indicates that **you're confident the work meets or exceeds Microsoft's standards for publication**, and will trigger the review process.</li><li>Your PR may be selected for initial review by the PnP content team. Following PnP review, you may receive questions or requests for additional changes. You should have initial feedback from PnP review within a few business days. If you have an urgent request or need to contact the team, please mention `@MicrosoftDocs/patterns-and-practices-team-pr-reviewers` in your PR and someone will get back to you. After the Azure patterns & practices content review is complete, a `PnP #sign-off` will be added.</li><li>Final review/merge is done by the PR review team. The PR team may also respond with feedback, categorized as "Blocking" (requires action from you), or "Non-blocking" (to be addressed in a future PR).</li></ol> |
| **Additional resources** | <ul><li>[learn.microsoft.com internal contributor guide](https://review.learn.microsoft.com/help/contribute/?branch=main)</li><li>Authoring templates: [architecture-center-pr](https://review.learn.microsoft.com/help/contribute/architecture-center/templates/sample-solution-templates?branch=main), [well-architected-pr](https://review.learn.microsoft.com/help/contribute/global-waf-template?branch=main)</li><li>To contact the Azure patterns & practices team use [e-mail](mailto:cact-pr-reviewers@microsoft.com?subject=Help%20with%20pull%20request), or @mention our GitHub team in your PR comments using: `@MicrosoftDocs/patterns-and-practices-team-pr-reviewers`</li></ul>
</details>

### CODE REVIEW

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:33]:
```suggestion
Sometimes, the event producer requires an immediate response from the event consumer, such as obtaining a customer eligibility before proceeding with an order. In event-driven architecture, synchronous communication can be implemented in two main ways:
```

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:35]:
```suggestion
- **Create two distinct queues**; a request queue and a reply queue. The event producer sends an asynchronous request to the request queue, pauses other operations, and waits for a response from the reply queue. Event consumers then process the request and send the reply back through the reply queue. This technique is a common implementation. The approach utilizes a session ID for tracking.
```

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:37]: I don't understand this one.  Let's keep it simple, and use Azure Storage queues as the example.  How would consumers know about this new queue to start monitoring?  Who deletes the queue (the consumer?  The publisher?).  I think this is too vague, and you're going to need to say a bit more.

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:35]: In retrospect I don't like where this in the document. Can you move this into the "challenges" section?  Meaning, a challenge is trying to use EDA for steps that need to follow a request-response pattern.  And this is the primary solution.  Also, let's make sure we're calling this is "request-response" implementation.  That is typically how we think about using buses for these situations.  Please also include a quick mention that some busses support this through a formal [ReplyTo](https://learn.microsoft.com/dotnet/api/azure.messaging.servicebus.servicebusmessage.replyto) property so that the sender directs the consumer on where to respond.

**Review by ckittel — CHANGES_REQUESTED** [review]

**Comment by azarboon** [comment]
@ckittel thanks for raising those fair points. I tried to incorporate them. Let me know if there is anything else needed.

**Comment by azarboon** [comment]
@ckittel  as always, it's pleasure to work with you

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:70]:
```suggestion
- Implementing a traditional request-response messaging pattern.
```

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:72]:
```suggestion
  Sometimes, the event producer requires an immediate response from the event consumer, such as obtaining a customer eligibility before proceeding with an order. In event-driven architecture, synchronous communication can be achieved through [request-response messaging](https://www.enterpriseintegrationpatterns.com/patterns/messaging/RequestReply.html). This pattern can be implemented in through utilizing multiple queues - a request queue, and a response queue.
```

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:74]:
```suggestion
  - **Establish two permanent queues**; a request queue and a reply queue. The event producer sends an asynchronous request to the request queue, pauses other operation on that task, and waits for a response in the reply queue. Event consumers then process the request and send the reply back through the preestablished reply queue. The approach utilizes a session ID for tracking so the event producer knows which message in the response queue is related to the specific request. This technique is a common implementation.
```

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:76]:
```suggestion
  - **Create an ephemeral queue for each response.** This version behaves like the prior solution, but instead of having a preexisting response queue established, this approach uses an ephemeral queue that is exclusively recognized by the event producer for a particular request. The event producer can either create this ephemeral queue or it may be generated automatically by the message broker. The event producer sends the initial request to the queue, specifying the name of the ephemeral queue in a [reply-to header](/dotnet/api/azure.messaging.servicebus.servicebusmessage.replyto) or another mutually agreed-upon custom attribute. Any messages directed to this queue are expected to be solely for the event producer that initiated them. Once the event consumer receives the message, processes the request, and sends a response to the reply queue indicated in the message header. The original event producer retrieves this response and subsequently deletes the ephemeral queue. This approach can lack scalability because a dedicated queue must be created and deleted for each request. Also, this solution could run into service limits on queue creation, rate or quantity.
```

**Review by ckittel — CHANGES_REQUESTED** [review]
How about this?

**Review by ckittel — APPROVED** [review]
@ShannonLeavitt, can you catch this one at some point for an editorial pass.  Can be before or after it merges.

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [reply docs/guide/architecture-styles/event-driven-content.md:72]: It does not contradict it, IMO.  The pattern is a queue for a request and a queue for a response. That doesn't imply anything about the lifecycle of either queue.

File: docs/guide/architecture-styles/event-driven-content.md
Comment (azarboon) [inline docs/guide/architecture-styles/event-driven-content.md:74]: > pauses other operation on that task

Just to be clear: My understanding is that event producer pauses all other operations, not just operations regarding that specific task.

**Comment by ckittel** [comment]
PnP #sign-off

#remove-label:"pnp-review-in-progress"
#remove-label:"do-not-merge"
#label:"ready-to-merge"

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [reply docs/guide/architecture-styles/event-driven-content.md:74]: Why would it do that?  That seems highly inefficient and would be a terrible bottleneck.

File: docs/guide/architecture-styles/event-driven-content.md
Comment (azarboon) [inline docs/guide/architecture-styles/event-driven-content.md:76]: > Create an ephemeral queue for each response.

My understanding is that the temporary queue is used for both request and reply. So, I think such title can be a bit misleading / confusing. But I leave it up to you.

Comment (ckittel) [reply docs/guide/architecture-styles/event-driven-content.md:76]: How would the consumer know about the queue?

Comment (azarboon) [reply docs/guide/architecture-styles/event-driven-content.md:76]: Fair point. Probably it has to be implemented somewhere code, but probably it can vary depending on the message broker. If you don't like this 2nd option, we can just remove it. I'm not religious about it :)

File: docs/guide/architecture-styles/event-driven-content.md
Comment (azarboon) [reply docs/guide/architecture-styles/event-driven-content.md:74]: Fair point.

File: docs/guide/architecture-styles/event-driven-content.md
Comment (azarboon) [reply docs/guide/architecture-styles/event-driven-content.md:76]: I'm going to remove the second part.

**Comment by ckittel** [comment]
#hold-off

**Comment by azarboon** [comment]
@ckittel as per our conversation, I removed the second implementation. I think it should be fine now but a final check won't hurt :)

File: docs/guide/architecture-styles/event-driven-content.md
Comment (ckittel) [inline docs/guide/architecture-styles/event-driven-content.md:72]:
```suggestion
  Sometimes, the event producer requires an immediate response from the event consumer, such as obtaining a customer eligibility before proceeding with an order. In event-driven architecture, synchronous communication can be achieved through [request-response messaging](https://www.enterpriseintegrationpatterns.com/patterns/messaging/RequestReply.html).

  This pattern is usually implemented by utilizing multiple queues - a request queue and a response queue. The event producer sends an asynchronous request to a request queue, pauses other operation on that task, and awaits a response in the reply queue; effectively turning this into a synchronous process. Event consumers then process the request and send the reply back through a response queue. This approach usually utilizes a session ID for tracking, so the event producer knows which message in the response queue is related to the specific request. The original request could also specify the name of the response queue, potentially ephemeral, in a [reply-to header](/dotnet/api/azure.messaging.servicebus.servicebusmessage.replyto) or another mutually agreed-upon custom attribute.
```

**Review by ckittel — CHANGES_REQUESTED** [review]

**Review by ShannonLeavitt — APPROVED** [review]

**Review by ckittel — APPROVED** [review]

**Pull request merged**
